vllm/spec_decode/ngram_worker.py

```diff
@@ -0,0 +1,260 @@
+import copy
+from typing import Dict, List, Optional
+
+import numpy as np
+import torch
+
+from vllm.sequence import SamplerOutput, SequenceGroupMetadata
+from vllm.spec_decode.interfaces import SpeculativeProposals
+from vllm.spec_decode.top1_proposer import DraftModelTop1Proposer
+from vllm.worker.worker import Worker
+
+
+class NGramWorker(Worker):
+    """NGramWorker provides a light drafter without need for model.
+
+    Current NGramWorker only implement prompt lookup decoding,
+    and in future we may also do RAG type drafter and other scenerios
+    which don't rely on LLM model to give proposals.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # Lazy initialization list.
+        self._proposer: DraftModelTop1Proposer
+
+    def fake_load_model(self):
+        pass
+
+    def fake_execute_model(
+        self,
+        seq_group_metadata_list: Optional[List[SequenceGroupMetadata]] = None,
+        blocks_to_swap_in: Optional[Dict[int, int]] = None,
+        blocks_to_swap_out: Optional[Dict[int, int]] = None,
+        blocks_to_copy: Optional[Dict[int, List[int]]] = None,
+        num_lookahead_slots: int = 0,
+    ):
+        pass
+
+    def set_ngram_window_size(self, ngram_prompt_lookup_max: int,
+                              ngram_prompt_lookup_min: int):
+        self.ngram_prompt_lookup_max = ngram_prompt_lookup_max
+        self.ngram_prompt_lookup_min = ngram_prompt_lookup_min
+
+    def init_device(self):
+        super().init_device()
+        # Since ngram worker use no model, fake load_model/execute_model
+        self.load_model = self.fake_load_model
+        self.execute_model = self.fake_execute_model
+
+        self._proposer = DraftModelTop1Proposer(
+            self,
+            self.device,
+            self.max_model_len,
+            self.vocab_size,
+        )
+
+    @torch.inference_mode()
+    def execute_model_multi_step(
```

Comment:
The current implementation couples the NGramWorker with the draft model worker a lot. In the future we will have RAG speculation, Medusa, Eagle, that all can use top-1 proposals. Can you look at how we can decouple Top1Proposal logic from both the draft model and ngram, so we can use it for the other ones in the future?

It will also simplify the ngram worker implementation; we won't need this notion of `execute_model_multi_step` and its baggage (shallow copy inputs, check KV size, append new tokens).

Reply:
I'm happy to meet to discuss this item further -- it will simplify things a lot in the future (cc @comaniac ).

Reply:
I currently move top-1 proposals into top1_proposer.py, and move shallow_copy_inputs into util.py, so that there is no [check KV size, append new tokens] logic for ngram now. 

For shallow_copy_inputs itself, it seems we may remove it from ngram, as it is true we don't need to do the append token stuff.

Reply:
+1 to removing shallow_copy_inputs / check KV size / append new tokens

> I currently move top-1 proposals into top1_proposer.py

The issue is the top1 proposer is still couples with the draft model -- it calls `execute_model_multi_step` and also relies on `max_model_len`. Top1Proposing should work for all proposal methods (ngram/eagle/medusa/draft), so before we can call your factor `Top1Proposer` we need to move out the draft-model specific components. As you have it, it's still a `DraftModelTop1Proposer`.